### PR TITLE
ENH: Added `bitwise_count` UFuncs

### LINKED ISSUE #16325: ENH: Expose `bit_count` ufunc equivalent to the scalar methods
Modern CPUs have a builtin [popcount](https://en.wikichip.org/wiki/population_count) instruction that counts the number of 1 bit in an unsigned integer. It would be great if numpy has a routine that exposes this operation.

Both GCC and MSVC have builtin functions for this instruction.

Possible usage:
```python
>>> np.popcount(4)
1
>>> np.popcount(np.arange(5, dtype=np.uint))
array([0, 1, 1, 2, 1])
```

### PR DESCRIPTION
**Pull request by ganesh-k13**
### `bitwise_count` UFuncs

Expose `bit_count` as an UFunc.
```py
>>> a = (2**np.arange(20) - 1)
>>> a
array([     0,      1,      3,      7,     15,     31,     63,    127,
          255,    511,   1023,   2047,   4095,   8191,  16383,  32767,
        65535, 131071, 262143, 524287])
>>> np.bitwise_count(a)
array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19])
```

This is a raw port of #19355, specifically from 600947eadf136c452e0fa0d8e9d17bf3caf180ab to 4ee8a40fc015916061b1441d4228b61eba7779df

#### Side Note:
Anyone new can use this a template to add new UFuncs (I guess). Related to #19494

TODO:
- [x] Figure out if it's up to date with latest main
-  [x] Add bencmarks (at least now :) )
-  [x] Commits are a mess, fix it
-  [x] Figure out the return dtype
-  [x] Clean `generate_umath`

Continues: #19355
Resolves: #16325

### CODE REVIEW

File: numpy/core/src/umath/funcs.inc.src [hunk 1]
```diff
@@ -267,6 +267,21 @@ npy_ObjectClip(PyObject *arr, PyObject *min, PyObject *max) {
     return o;
 }
 
+static PyObject *
+npy_ObjectPopCount(PyObject *obj) {
+    PyObject *result = NULL;
+
+    /* Try to use inbuilt popcount if available */
+    static PyObject *builtin_popcount_func = NULL;
+    builtin_popcount_func = PyObject_GetAttrString(obj, "bit_count");
```
Comment (seberg) [inline numpy/core/src/umath/funcs.inc.src:276]: Hmm, most ufuncs already try to call a method of the same name when it exists?  There should be a shorthand to do this in `generate_umath` without the need to define this function.

**Review by seberg — COMMENTED** [review]
Cool!  One small comment for simplifying a bit, but the code looks great on first sight!

My two main questions are:
* Did we decide what the output dtype is?  The default integer, the same integer?
* While I am in favor of adding the `bit_count` ufunc (which still may need more discussion unfortunately), I would still need a bit convincing on adding a method, it doesn't seem as clear to me.  There could be an argument of _where_ to add the ufunc, but hopefully we don't lose track of this because of such questions.

**Comment by ganesh-k13** [comment]
Thanks for the review @seberg

> Did we decide what the output dtype is? The default integer, the same integer?
For the scalar version we decided to always output an 8 bit (`uint8_t`). Although here it might be the case, will need to change it.

>  There could be an argument of where
This brings a very good point on if we can create a `np.bits` namespace of sorts and club the bits stuff there, what do you think?

File: numpy/core/src/umath/funcs.inc.src (hunk 1 above)
Comment (ganesh-k13) [reply numpy/core/src/umath/funcs.inc.src:276]: ohh yeah, let me see if we can auto-gen this, will make things easier.

Comment (ganesh-k13) [reply numpy/core/src/umath/funcs.inc.src:276]: Ohh also makes sense now, why not all umath functions are here.

**Comment by softwaredoug** [comment]
This PR seems so close. Anything I can do to help get this over the finish line @ganesh-k13? For certain similarity functions (like hamming distance) the bit count function is the bottleneck, and this could really unlock pure python solutions without needing to resort to something in a lower level language.

**Comment by ganesh-k13** [comment]
Hey @softwaredoug , we support `bit_count` today and was added via #19355. This PR is mainly to get a `bit_count` UFunc so we can operate it on arrays, etc.

Well to answer your question on moving it forward, implementation wise, perhaps the `generate_umath` can be made simpler. But the bigger hurdle would be convince the mailing list that this function is needed and doesn't add more to the already bloated namespace.

So in conclusion, if you feel this has a genuine use case for many in the community (the UFunc of course, the scalar version is already in), you can start a thread in the mailing list. Now I can prioritize this to completion if everyone is ok with adding this UFunc and if you want to contribute, I can help you in any way with terms of navigating this bit.

**Comment by seberg** [comment]
I will give my :+1: for the ufunc, but not for the method.  Feel free to announce to the mailing list that pending no substantial objects we will move forward with adding it.

The return dtype has to be resolved, but I guess it was defined as uint8 for the scalars?  In which case we have prior art.

**Comment by softwaredoug** [comment]
Awesome - thanks @ganesh-k13 and @seberg 

Maybe this is my dumpy ignorance, but is there a way to efficiently apply the scalar function to a whole array? Without something like `apply_along_axis` which in my mental model wouldn't be particularly fasts?

**Comment by seberg** [comment]
No, there isn't an efficient way (maybe a JIT'er like `numba` can do it, not sure).  You could write your own ufunc, its not actually that terribly hard, but the learning curve is just quite steep, especially if you are not used to C ([ufunclab](https://github.com/WarrenWeckesser/ufunclab) is my goto example for external ufuncs currently).

**Comment by ganesh-k13** [comment]
Hey @BvB93 , basic question: If we fix the return type of UFunc to one dtype, where do we make the typing changes? 

Now I do get `ValueError: Unexpected reveal line format: 23: error: Name "ctypes._Pointer" is not defined` in numpy/typing/tests/test_typing.py, but I don't know how to interpret this. I guess we need to make some change in numpy/typing/tests/data/reveal/ufuncs.pyi?

File: numpy/core/src/umath/funcs.inc.src (hunk 1 above)
Comment (ganesh-k13) [reply numpy/core/src/umath/funcs.inc.src:276]: Hey @seberg , I dug a bit deeper here and it seems all the `o` supported dtypes have a function in `funcs.inc.srs` and it does not seem automatic. Am I missing something here?

Comment (seberg) [reply numpy/core/src/umath/funcs.inc.src:276]: Ah, there is the `"P"` character code that means: "Object, but just call the method".

Comment (ganesh-k13) [reply numpy/core/src/umath/funcs.inc.src]: Ah nice, I missed it. Thanks! Added via [b86a322](https://github.com/numpy/numpy/pull/21429/commits/b86a322d45950469f3d84e5c05972212192e6ed0)

**Comment by ganesh-k13** [comment]
Updated `bit_count` (just the UFunc) to `bitwise_count`. Reference to discussion regarding the same: [discussions](https://mail.python.org/archives/list/numpy-discussion@python.org/message/VUKAKP7MIJPJSRGAJPZXKBOLGOUVGPG7/)

**Review by seberg — COMMENTED** [review]
Thanks.  Lots of nitpicks of the optional kind.  I have to think about the naming a bit more.
Overall, I lean against adding any methods, which is the only meaningful change besides possible name bike-shedding.

File: numpy/core/src/multiarray/scalartypes.c.src [hunk 2]
```diff
@@ -2192,6 +2192,9 @@ static PyMethodDef gentype_methods[] = {
     {"sum",
         (PyCFunction)gentype_sum,
         METH_VARARGS | METH_KEYWORDS, NULL},
+    {"bit_count",
+        (PyCFunction)gentype_bit_count,
+        METH_VARARGS | METH_KEYWORDS, NULL},
```
[the remark this answers is not in the document]
Comment (ganesh-k13) [reply numpy/core/src/multiarray/scalartypes.c.src:2197]: Ah, I'm actually not even sure why I added it to begin with :), wasn't part of the plan. Yeah I'll remove it and if needed, I'll add it back

File: numpy/core/code_generators/ufunc_docstrings.py [hunk 3]
```diff
@@ -4214,3 +4214,38 @@ def add_newdoc(place, name, doc):
...
+    ----------
+    x : array_like, unsigned int
+        Input array.
+    $PARAMS
+
+    Returns
+    -------
+    y : ndarray
+        The corresponding number of 1-bits in the input.
+        $OUT_SCALAR_1
```
[the remark this answers is not in the document]
Comment (ganesh-k13) [reply numpy/core/code_generators/ufunc_docstrings.py:4233]: Ah yes, thanks, I forgot to add it.

File: numpy/core/src/multiarray/methods.c [hunk 4]
```diff
@@ -3076,9 +3082,11 @@ NPY_NO_EXPORT PyMethodDef array_methods[] = {
     {"__dlpack__",
         (PyCFunction)array_dlpack,
         METH_FASTCALL | METH_KEYWORDS, NULL},
-
     {"__dlpack_device__",
         (PyCFunction)array_dlpack_device,
         METH_NOARGS, NULL},
+    {"bitwise_count",
+        (PyCFunction)array_bitwise_count,
+        METH_VARARGS | METH_KEYWORDS, NULL},
```
[the remark this answers is not in the document]
Comment (ganesh-k13) [reply numpy/core/src/multiarray/methods.c:3090]: This might be handy? If we want to support it on other sequences? I'm guessing there will be some other way though without this addition? happy to do both

[the remark this answers is not in the document]
Comment (ganesh-k13) [reply numpy/core/src/multiarray/methods.c]: I removed it via [d7f364f](https://github.com/numpy/numpy/pull/21429/commits/d7f364fee9d1a85b08126ff723f8d884f801e8ad). I'm a bit confused now, I thought this adds `bitwise_count` to np namespace?

File: numpy/core/src/multiarray/scalartypes.c.src (hunk 2 above)
[the remark this answers is not in the document]
Comment (ganesh-k13) [reply numpy/core/src/multiarray/scalartypes.c.src]: I removed it via [d7f364f](https://github.com/numpy/numpy/pull/21429/commits/d7f364fee9d1a85b08126ff723f8d884f801e8ad)

File: numpy/core/code_generators/ufunc_docstrings.py [hunk 5]
```diff
@@ -4214,3 +4214,38 @@ def add_newdoc(place, name, doc):
...
+
+    References
+    ----------
+    .. [1] https://stackoverflow.com/a/109025/5671364
+
+    .. [2] Wikipedia, "Hamming weight",
+           https://en.wikipedia.org/wiki/Hamming_weight
```
[the remark this answers is not in the document]
Comment (ganesh-k13) [reply numpy/core/code_generators/ufunc_docstrings.py:4241]: Added some direct references through [0ff8230](https://github.com/numpy/numpy/pull/21429/commits/0ff8230fc506e7cefa018f43089a262db7249695)

File: numpy/core/code_generators/ufunc_docstrings.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ganesh-k13) [reply numpy/core/code_generators/ufunc_docstrings.py:4234]: added in [0ff8230](https://github.com/numpy/numpy/pull/21429/commits/0ff8230fc506e7cefa018f43089a262db7249695)

File: doc/release/upcoming_changes/19355.new_feature.rst [hunk 6]
```diff
@@ -0,0 +1,14 @@
+`np.bitwise_count` to compute the number of 1-bits in an integer
+----------------------------------------------------------------
+
+This new function counts the number of 1-bits in a number.
+These work on all the numpy integer types, as well as the
+builtin arbitrary-precision `Decimal` and `long` types.
```
[the remark this answers is not in the document]
Comment (ganesh-k13) [reply doc/release/upcoming_changes/19355.new_feature.rst]: removed in [0ff8230](https://github.com/numpy/numpy/pull/21429/commits/0ff8230fc506e7cefa018f43089a262db7249695)

File: doc/release/upcoming_changes/19355.new_feature.rst [hunk 7]
```diff
@@ -0,0 +1,14 @@
+`np.bitwise_count` to compute the number of 1-bits in an integer
+----------------------------------------------------------------
+
+This new function counts the number of 1-bits in a number.
```
[the remark this answers is not in the document]
Comment (ganesh-k13) [reply doc/release/upcoming_changes/19355.new_feature.rst:4]: Added in [0ff8230](https://github.com/numpy/numpy/pull/21429/commits/0ff8230fc506e7cefa018f43089a262db7249695), I added a small `works on integer-like`, just to reference the fact we support object loops, is that ok?

File: doc/release/upcoming_changes/19355.new_feature.rst [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+These work on all the numpy integer types, as well as the
+builtin arbitrary-precision `Decimal` and `long` types.
+Internally, this function calls `np.bit_count`
```
[the remark this answers is not in the document]
Comment (ganesh-k13) [reply doc/release/upcoming_changes/19355.new_feature.rst]: removed in [0ff8230](https://github.com/numpy/numpy/pull/21429/commits/0ff8230fc506e7cefa018f43089a262db7249695)

**Comment by ganesh-k13** [comment]
Resolved merge conflicts (rebased with latest main)

**Comment by ganesh-k13** [comment]
Long overdue look at the Array API standard. Two callouts:

1. [elementwise_functions](https://data-apis.org/array-api/2022.12/API_specification/elementwise_functions.html) section does not include `bitwise_count`, but a note: All other functions that operates on bits use `bitwise_*`
2. [bitwise_operators](https://data-apis.org/array-api/2022.12/API_specification/array_object.html#bitwise-operators) specification only talks of: `Bitwise operators should be defined for arrays having integer and boolean data types.`, which we do follow in this change list.

File: numpy/core/code_generators/ufunc_docstrings.py [hunk 9]
```diff
@@ -4214,3 +4214,41 @@ def add_newdoc(place, name, doc):
     array([ 0, 20, 20, 60, 20, 20])
 
     """)
+
+add_newdoc('numpy.core.umath', 'bitwise_count',
```
Comment (ganesh-k13) [inline numpy/core/code_generators/ufunc_docstrings.py:4218]: For some reason, this is not getting picked. Checking.

Comment (ganesh-k13) [reply numpy/core/code_generators/ufunc_docstrings.py:4218]: NVM, issue was in loops.c.src:444

**Comment by ganesh-k13** [comment]
Are any more changes needed here @seberg?

**Comment by softwaredoug** [comment]
Just checking in on this fastastic work! 

A note I rebased and fixed the linting errors in this branch, if anyone wants to cherry-pick the commit. 

https://github.com/softwaredoug/numpy/tree/enh_16325_popcount_ufunc

**Comment by softwaredoug** [comment]
Not sure if expected, I also had to add an bitwise_count in __init__.py for this to work

Once I did this, I see a 10x improvement in counting bits compared to the naive / non popcount solution with lots of bittwiddling 🎉 

https://gist.github.com/softwaredoug/8210fa941f8b806b9eea83fd1e762d03

**Comment by softwaredoug** [comment]
> We use the same parallel counting implementation under the hood.

Ah interesting! I had assumed we used some compiler intrinsics that accessed the CPU's popcount instruction (at least where possible).

**Comment by ganesh-k13** [comment]
Ah yes, we first check if the system has the instructions and only use this as a last resort. You can find them a few lines below the above implementation.

**Comment by rgommers** [comment]
@ganesh-k13 CI is kind of unhappy - but that looks easy to resolve.

@seberg this PR came up in the community meeting yesterday. It looks like it's in good shape, and adding `np.bitwise_count` seems like a go. I resolved a number of your comments that were addressed; I left a few that I wasn't immediately sure about open. It'd be nice to merge this soon. Would you like to have another look or are you happy as is?

**Comment by seberg** [comment]
I always would have liked more than us two voice an opinion about the name.  I don't particularly love `bitwise` because we create the dissonance with the method (which probably even breaks object dtype).
But maybe someon else from yesterday can explicitly voice their opinion.

**Comment by seberg** [comment]
Otherwise, I will try to have a pass on friday, but if not, don't wait for me: I will be on vacation for a while and probably not check in much.

**Review by seberg — COMMENTED** [review]
LGTM modulo bikeshed as expected.

**Comment by charris** [comment]
I would have preferred something like `count_bits`, if that is what it does, but given the `bitwise*` precedent we might as well stick with that.

**Comment by seberg** [comment]
Yeah, to be fair neither is perfect, right.  Even sum might be better.  `bitwise` for me fits better to binary operations: they operate bitwise on two numbers.  It doesn't seem quite fitting for a single number to me.  `left_shift` would be a precendence to me, but unfortunately it doesn't use a clear name :).

**Comment by ganesh-k13** [comment]
The failures in CI might have been due to some changes in https://github.com/numpy/numpy/issues/24507. I'm tagging the issue just in case. If the import added in [7920eff](https://github.com/numpy/numpy/pull/21429/commits/7920eff539f824895922608d794c840c2c7cc6fc) is not as planned, please let me know.

**Comment by seiko2plus** [comment]
I have made few changes to the code to allow auto-vectorization, and here the result on x86 with avx2 enabled:

```Bash
CPU:
❯ lscpu
Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         39 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  8
  On-line CPU(s) list:   0-7
Vendor ID:               GenuineIntel
  Model name:            Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
  
❯ spin bench --compare enh_16325_popcount_ufunc_before -t bitwise_count
| Change   | Before [03fc9291] <enh_16325_popcount_ufunc_before>   | After [fc6b5911] <enh_16325_popcount_ufunc>   |   Ratio | Benchmark (Parameter)                                                             |
|----------|-------------------------------------------------------|-----------------------------------------------|---------|-----------------------------------------------------------------------------------|
| -        | 253±0.1μs                                             | 170±0.1μs                                     |    0.67 | bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'bitwise_count'>, 1, 1, 'H') |
| -        | 180±0.4μs                                             | 58.9±0.6μs                                    |    0.33 | bench_ufunc.UFunc.time_ufunc_types('bitwise_count')                               |
| -        | 353±0.8μs                                             | 117±1μs                                       |    0.33 | bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'bitwise_count'>, 1, 1, 'h') |
| -        | 564±0.5μs                                             | 121±0.2μs                                     |    0.22 | bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'bitwise_count'>, 1, 1, 'l') |
| -        | 566±4μs                                               | 123±0.4μs                                     |    0.22 | bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'bitwise_count'>, 1, 1, 'q') |
| -        | 621±3μs                                               | 108±1μs                                       |    0.17 | bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'bitwise_count'>, 1, 1, 'i') |
| -        | 562±0.5μs                                             | 79.6±0.2μs                                    |    0.14 | bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'bitwise_count'>, 1, 1, 'L') |
| -        | 562±0.4μs                                             | 80.5±0.2μs                                    |    0.14 | bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'bitwise_count'>, 1, 1, 'Q') |
| -        | 679±3μs                                               | 75.1±0.06μs                                   |    0.11 | bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'bitwise_count'>, 1, 1, 'I') |
| -        | 269±1μs                                               | 11.9±0.08μs                                   |    0.04 | bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'bitwise_count'>, 1, 1, 'B') |
| -        | 309±0.2μs                                             | 12.8±0.05μs                                   |    0.04 | bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'bitwise_count'>, 1, 1, 'b') |
```

**Comment by ganesh-k13** [comment]
Wow! Thanks Sayed!

**Comment by seiko2plus** [comment]
> Wow! Thanks Sayed!

Not that good, the GCC compiler was expected to apply something similar to Harley-Seal popcount algorithm for optimized machine code generation. However, it defaulted to using the standard popcnt instruction without additional optimizations. To potentially enhance performance, we could improve the existing "fast macros" to provides better hints for the compiler maybe by manually unrolling specialized loops counting on C++ meta or/and implementing the algorithm with handwritten generic SIMD instructions.

File: numpy/typing/tests/data/reveal/ufuncs.pyi [hunk 10]
```diff
@@ -74,3 +75,14 @@ assert_type(np.matmul.signature, Literal["(n?,k),(k,m?)->(n?,m?)"])
...
+assert_type(np.bitwise_count.__name__, Literal['bitwise_count'])
+assert_type(np.bitwise_count.ntypes, Literal[11])
+assert_type(np.bitwise_count.identity, None)
+assert_type(np.bitwise_count.nin, Literal[1])
+assert_type(np.bitwise_count.nout, Literal[1])
+assert_type(np.bitwise_count.nargs, Literal[2])
+assert_type(np.bitwise_count.signature, None)
+assert_type(np.bitwise_count.identity, None)
+assert_type(np.bitwise_count(i8), Any)
+assert_type(np.bitwise_count(AR_i8), Any)
```
Comment (ganesh-k13) [inline numpy/typing/tests/data/reveal/ufuncs.pyi:88]: This is failing now. I'll take a look tomorrow.

**Comment by ganesh-k13** [comment]
@seberg I've resolved the conflicts and UT failures, let me know if anything else needed here.

File: numpy/typing/tests/data/reveal/ufuncs.pyi (hunk 10 above)
Comment (ganesh-k13) [reply numpy/typing/tests/data/reveal/ufuncs.pyi]: I think `npt.NDArray[Any]` is correct?

Comment (BvB93) [reply numpy/typing/tests/data/reveal/ufuncs.pyi]: Correct, I'm somewhat surprised that it even produced an `Any` here at one point. Though it could have been accepted due to the somewhat broad substring matching used previously.

Comment (ganesh-k13) [reply numpy/typing/tests/data/reveal/ufuncs.pyi]: Ah I see. That makes sense why it was passing before.

**Comment by ganesh-k13** [comment]
@seberg any other changes needed here?

**Comment by seberg** [comment]
Sorry, do you have time to resolve the conflict, I think that was really all that was left here, just I was in and out a bit much.

**Comment by ganesh-k13** [comment]
@seberg it seems to pass now 👍 , good to go from my end.

**Comment by seberg** [comment]
Thanks, let's get it in before it diverges again.

**Comment by ganesh-k13** [comment]
This is probably the longest I've worked on a PR :). Now time to get hex/fromhex in 😉

**Comment by jakevdp** [comment]
This is great! One observation – the implementation does not appear to support general output dtypes:
```python
>>> import numpy as np
>>> np.bitwise_count(2, dtype='int32')
# ---------------------------------------------------------------------------
# TypeError                                 Traceback (most recent call last)
# <ipython-input-6-f7427f7dde09> in <cell line: 2>()
#       1 import numpy as np
# ----> 2 np.bitwise_count(2, dtype='int32')

# TypeError: No loop matching the specified signature and casting was found for ufunc bitwise_count
```
It would be nice if we could support the `dtype` argument.
